src/login_nopam.c

```diff
@@ -265,33 +263,18 @@ static bool user_match (const char *tok, const char *string)
 
 static const char *resolve_hostname (const char *string)
 {
-	int                  gai;
-	char                 *p;
-	struct addrinfo      *addrs;
-	struct sockaddr      *sa;
-	struct sockaddr_in   *sin;
-	struct sockaddr_in6  *sin6;
+	int              gai;
+	char             *p;
+	struct addrinfo  *addrs;
 
 	gai = getaddrinfo(string, NULL, NULL, &addrs);
 	if (gai != 0)
 		goto notfound;
 
-	sa = addrs[0].ai_addr;
-	switch (addrs[0].ai_addr->sa_family) {
-	case AF_INET:
-		sin = (struct sockaddr_in *) sa;
-		inet_ntop(AF_INET, &sin->sin_addr,
-		          inet_ntop_buffer, NITEMS(inet_ntop_buffer));
-		p = inet_ntop_buffer;
-		break;
-	case AF_INET6:
-		sin6 = (struct sockaddr_in6 *) sa;
-		inet_ntop(AF_INET6, &sin6->sin6_addr,
-		          inet_ntop_buffer, NITEMS(inet_ntop_buffer));
-		p = inet_ntop_buffer;
-		break;
-	default:
-		p = "Hypothetical future inet version?";
+	p = inet_sockaddr2str(addrs[0].ai_addr);
+	if (p == NULL) {
+		SYSLOG ((LOG_ERR, "Hypothetical future IPv7?"));
+		abort();
```

Comment:
I assume this is a WIP, but abort()ing here is a change in behavior.

Reply:
No, it's not a WIP.

Yes, this is an important change in behavior, which is why I kept it separate from the cleanups for POSIX.1-2008:
<https://github.com/shadow-maint/shadow/pull/616#issuecomment-1362720236>.

The change in behavior (documented in the commit message) is:

The old gethostbyname(3) only supports IPv4.  The replacement, getaddrinfo(3) supports both IPv4 and IPv6 (and is not inherently limited to those, so a hypothetical IPv7 could also be supported in the future).  By changing the function, we're adding support for IPv6, at least in that part of the code.

Because the function is not inherently limited to IPv4/IPv6, I wanted to be cautious, and abort(3) on anything that's not 4 or 6.  So, if IPv7 is ever a thing, this code will gracefully abort (there's nothing better we can do), instead of doing harmful stuff.

Reply:
Although, I should have aborted in the first commit already.  I forgot to do that.  Will fix in a moment.

Reply:
Fixed.  Now I'm aborting since the first commit.

Reply:
Hm, well, "gracefully abort" is an interesting phrase, but I see, if a site has user and no from table entries, the login should proceed, so it's ok.

Thanks.